data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,675 @@
+
+In practice, curried functions are the default form functions should take because:
+- They allow partial application
+- Less editing, no parentheses or commas
+- No pattern matching over a tuple
+
+## Functions With Side-Effects
+
+With respect to its type, a function is processes input data from its *domain*, the set of input values, and produce a result data from its *codomain*, the set of output values.
+```ocaml
+# string_of_int;;
+- : int -> string = <fun>
+```
+The function `string_of_int`
+* Has domain `int`
+* Has codomain `string`
+
+However, some functions either take input data outside of their domain or produce data outside of their codomain. These out-of-signature data are called effects, or side effects. Input and output (I/O) are the most common forms of effects. Input is out-of-domain data and output is out-of-codomain data. However, the result of functions returning random numbers or the current time is influenced by external factors, which is also called an effect. The external factor is out-of-domain input. Similarly, any observable phenomena triggered by the computation of a function is out-of-codomain output.
+
+In practice, what is considered an effect is an engineering choice. In most circumstances, I/O operations are considered as effects, unless they are ignored. Electromagnetic radiation emitted by the processor when computing a function isn't usually considered a relevant side-effect, except in some security-sensitive contexts. In the OCaml community, as well as in the wider functional programming community, functions are often said to be either [pure](https://en.wikipedia.org/wiki/Pure_function) or impure. The former does not have side effects, the latter does. This distinction makes sense and is useful. Knowing what the effects are, and when are they taking place, is a key design consideration. However, it is important to remember this distinction always assumes some sort of context. Any computation has effects, and what is considered a relevant effect is a design choice.
+
+Since, by definition, effects lie outside function types, a function type can't reflect the effects a function may have. However, it is important to document the intended side effects a function may have. Consider the `Unix.time` function. It returns the number of seconds elapsed since January 1, 1970.
+```ocaml
+# Unix.time ;;
+- : unit -> float = <fun>
+```
+If you're getting an `Unbound module error` in macOS, run this first: `#require "unix";;`.
+
+To produce its result, no data needs to be passed to that function. The result is entirely determined by external factors. If it was passed information, it would not be used. But something must be passed as a parameter to trigger the request of the current time from the operating system.
```

Comment:
```suggestion
To produce its result, no data needs to be applied to that function. The result is entirely determined by external factors. If information was applied, it would not be used. But something must be passed as a parameter to trigger the request of the current time from the operating system.
```
Should we use "applied" here instead of "passed"? (Although, I think people will start to see the two mean the same thing.)

Reply:
Perfect!